### ISSUE TITLE
API Proposal: Expose global allocations akin to AppDomain.MonitoringTotalAllocatedMemorySize

### ISSUE THREAD

**Issue by NickCraver** [dotnet/corefx#30644]
This is a follow-up to #10157.

In full framework, we have [AppDomain.MonitoringTotalAllocatedMemorySize](https://msdn.microsoft.com/en-us/library/system.appdomain.monitoringtotalallocatedmemorysize), but no such equivalent exists in .NET Core. In #10157, there was at one point a proposal for:
```c#
namespace System
{
    public static class GC
    {
		public long GetAllocatedBytesForAllThreads();
    }
}
```
...but this was dropped due to no known scenarios. We now have that scenario commonly: monitoring allocations when working with `async`. This affects in general the ability for BenchmarkDotNet to function (see https://github.com/dotnet/BenchmarkDotNet/issues/723) and concretely affects our ability to benchmark Pipelines code (see https://github.com/mgravell/Pipelines.Sockets.Unofficial/pull/4).

Can we revisit this API addition, or alternatively, flesh out `AppDomain.MonitoringTotalAllocatedMemorySize` with the same behavior? I'm not sure which is better since the landscape has changed since the last proposal in closer-equivalency between platforms. The bottom line is: we can't get this information currently in Core and need a way to do so.

cc @terrajobst @adamsitnik @stephentoub @benaadams

**Comment by NickCraver** [comment]
cc @vancem @Maoni0

**Comment by Maoni0** [comment]
the async scenario has always existed and it was a TODO on my list. however I have not had time to work on it. but I wanted to clarify what that was supposed to work. 

```csharp
        static async Task DoStuff1()
        {
            var tracker = AllocationTracker.Start();

            var manualCount = await DoStuff2();

            Console.WriteLine($"Manual count: {manualCount}, Tracker count: {tracker.BytesAllocated}");
        }
```
and `tracker.BytesAllocated` would include what `DoStuff2` allocates. this is really the same as `GetAllocatedBytesForCurrentThread` but also taking `async` into consideration.

do you believe you actually want `GetAllocatedBytesForAllThreads` for all threads? what I meant by "I have not heard of scenarios" was for this API. do you prefer to have what I described above or actually a value that tells you about all threads in the process? the latter will not be accurate by nature - unless you actively stop all threads and call this API, by the time you get the value of course other threads could have already allocated a bunch of memory. 

if you could let me which one is actually your scenario it would be helpful.

**Comment by adamsitnik** [comment]
From my perspective, of a person who creates benchmarking tool I need an API that will give me the total number of allocated bytes. For all threads, no matter of processor affinity.

Because of the nature of existing AppDomain API for desktop .NET BenchmarkDotNet is already designed in a way that after running the benchmarks and measuring the time we EnableMontitoring in current AppDomain and perform single iteration (which consists of multiple benchmark invocations) to get the number of allocated bytes. Any time overhead is not an issue, because the timing result from this extra iteration has overhead and is not included in the final results. [Code](https://github.com/dotnet/BenchmarkDotNet/blob/master/src/BenchmarkDotNet/Engines/Engine.cs#L152-#L170)

what I mean by that is that even if we have to do some extra work to start/stop something it's perfectly fine. BenchmarkDotNet is going to hide this complexity from the end users, who are going to get the accurate number of allocated bytes. 

I have already ported entire CoreCLR benchmark suite to BenchmarkDotNet and now I am working on CoreFX and this feature is going to be used to track the regressions/improvements in allocations for our entire stack.

**Comment by NickCraver** [comment]
+1 to @adamsitnik's behavior - given the primary scenario is benchmarking the "overall" whether that overall is an entire application with a lot of features or a single benchmark snippet with 1 function and purpose.

@Maoni0 I agree the `async`-down tracking would be useful in itself, but IMO that's another API. It'd be interesting for doing a lot of custom benchmarking perhaps, but given all the other things a benchmark wants to test like cold start, spin up, and general before/after testing of memory, inlining, runtime, etc. I'm not sure a tool would change to using it vs. global.

The use case I'd actually have for such an API is not benchmarking locally, but helping track down jobs in production at scale, for example I could see MiniProfiler adding such a `using()` wrapper with an `AllocationTracker` to analyze memory (if enabled) to help find issues. I could see that being *extremely* useful, especially if such trackers could be nested, e.g. starting one in a child method and seeing where the allocations went in a tree-like form to see where that memory is going in a hurry. I'd *love* this API, but I'm not sure I could make a strong argument for it other than the profiling use case. The global one's a bit more pressing as it's affecting our ability to tune code today.

**Comment by Maoni0** [comment]
I see. thanks for the explanation.

if something like `AppDomain.MonitoringTotalAllocatedMemorySize` would suffice, that's pretty simple - the code is there, it's hanging off of AD which doesn't exist in core but you just need to have it record globally somewhere. you (or anyone else who's interested) are both perfectly welcome to make a change that gives the same info out as a different API (obviously go through the standard API review process with the API folks). there are just 2 places in GC that give memory out, if you search for `RecordAllocBytes` you'll see them. that's how the `MonitoringTotalAllocatedMemorySize` API gets its info.

What I should mention is this API is not completely accurate because it was for the whole AD usage and the accuracy it provides was sufficient whereas `GC.GetAllocatedBytesForCurrentThread` is perfectly accurate (well, if it's not it just means I have a bug :P). the inaccuracy of `MonitoringTotalAllocatedMemorySize` comes from the fact that when GC gives out an allocation context (which is 8k or smaller) the whole AC is reported even when it's not used up. so you could be in a situation where you keep getting these 8k bytes and only use 6k so you are overcounting 2k each time but that's certainly not common.

**Comment by jkotas** [comment]
> change that gives the same info out as a different API 

Why not just use the existing AppDomain APIs for it? They unconditionally throw in .NET Core today. It is fine to change them to do something useful.

**Comment by Maoni0** [comment]
> Why not just use the existing AppDomain APIs for it? They unconditionally throw in .NET Core today. It is fine to change them to do something useful.

because there's no ADs in coreclr and it would be weird to have an AD API? :P I don't have a strong opinion on this though.

**Comment by Maoni0** [comment]
of course if you want the accuracy of `GC.GetAllocatedBytesForCurrentThread` it's not that hard to do if you have a convenient place where you have all the threads stopped - you could do it by keeping track of the threads that died and remembering their allocated bytes; and enumerating the current threads to get their current allocated bytes. it's more work so obviously only do it if you really need it.

**Comment by jkotas** [comment]
> because there's no ADs in coreclr and it would be weird to have an AD API

There is public AppDomain type in .NET Core: https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Extensions/ref/System.Runtime.Extensions.cs#L24 . And these API do exist on it already for compatibility. They just always throw today.

**Comment by NickCraver** [comment]
If emulating the old API is preferable, +1 - awesome. As long as the data is available to use in the end, either one seems just fine. Does that remove the need for an API review, since we're just lighting up an existing API? If so I'll take a look at a PR here.

**Comment by stephentoub** [comment]
> Does that remove the need for an API review, since we're just lighting up an existing API?

Yup.

>  If so I'll take a look at a PR here.

Thanks.

**Comment by NickCraver** [comment]
@Maoni0 @stephentoub I looked at this a bit today, initial findings:
- This is pretty tied into app domains are they were/are on full framework builds of the GC
- `FEATURE_APPDOMAIN_RESOURCE_MONITORING` gates all of this, from both callers and tracking
  - Both `RecordAllocBytes` and `nGetTotalAllocatedMemorySize` are behind this define
- Exposing the API itself via `GCInterface` (are there any issues with additions to this? I assume not)
- We'd have to make GC changes here to bring some of the tracking that (as far as I can tell) aren't defined in the .NET Core GC build back in the fold
- I'm no C++ or GC expert - all of this may be wrong, just trying to help solve a problem. If I'm way off base, *please* correct me.

I don't see how this wouldn't have some allocating-time impact with the additional tracking not present in .NET Core today. For example:
```cpp
#ifdef FEATURE_APPDOMAIN_RESOURCE_MONITORING
    if (g_fEnableARM)
    {
        AppDomain* alloc_appdomain = GetAppDomain();
        alloc_appdomain->RecordAllocBytes (limit_size, heap_number);
    }
#endif //FEATURE_APPDOMAIN_RESOURCE_MONITORING
```
...which runs for (I think) every 8k allocation (@Maoni0 am I way off there?) Also, it obviously relies on `AppDomain*` itself, so that's a fundamental issue. It's tradeoff between compile-def switching implementations in the GC for fetching what is surfaced as the same API vs. a simpler implementation underneath (a la `GC.GetAllocatedBytesForCurrentThread`) that makes 1:1 with the path and sourcing all the way up the code tree...but is a new/different API on the consumption side.

From a code-in-/coreclr standpoint, a new API (as the original issue) would be far cleaner, but as @jkotas notes that's not necessarily the way to go on the API surface. As a user, my thoughts are: app domains don't exist in .NET Core anymore, so having to shift away from their APIs for things is nothing new or isolated. I wouldn't mind the API exposure in either place...people are already forking today.

I'm not sure I'd be comfortable making the change to "pretend" to be app-domain underneath given my level of expertise in this area (though a new API seems doable) - either way, I want to help this issue along as much as possible if this is something we can do/fix. Thoughts?

**Comment by jkotas** [comment]
The internal implementation should be simple either way. Yes, there needs to be a new API on GC/EE interface for it. The implementation should not be based on `FEATURE_APPDOMAIN_RESOURCE_MONITORING`. As you have pointed out, it had unnecessary overhead.

If there is a significant benefit in exposing this via a new (duplicate) public API in a better place, I would be fine with me. However, even if we do that - the existing AppDomain API should be hooked up to return the same value in .NET Core for better compatibility. It costs next to nothing to do and the API is not going to ever go away because of it is part of netstandard20.

**Comment by Maoni0** [comment]
yeh, as @jkotas pointed out, you can have the same AD API but just give it a different implementation. that's what I meant by " it's hanging off of AD which doesn't exist in core but you just need to keep that info somewhere else". I have a [commit](https://github.com/Maoni0/coreclr/commit/25035e93bdd7effb54c30e85d32854f7f9052927) that illustrates the "keeping it somewhere else" part (ie, as a per heap member on gc_heap), please feel free to build on this. note that this doesn't build because I added a new `GCHeap::GetTotalAllocatedBytes` but didn't add the declaration or anything, only the implementation. this will need to be exposed on the GC/EE interface and the AD API can then call it.

**Comment by vancem** [comment]
I think for the use case that is being described, the inaccuracy of the AppDomain-wide implementation is reasonably severe (they probably wanted to track regressions < 8K accurately without resorting to multiple iterations).     Given that there is an 'obvious' implementation that does not have this limitation (iterate over all managed threads and sum the per-thread stats), we should consider implementing it that way (still very straightforward).

**Comment by Maoni0** [comment]
> iterate over all managed threads and sum the per-thread stats

and track the dead ones. as I mentioned above, this is not hard to do; it's just a matter of whether it's needed. I'll let folks that wanted this decide whether it's needed...

**Comment by NickCraver** [comment]
I agree the more accurate the better - that's the primary use case (at least for us). However I'm not sure how the proposed behavior works with thread pooling in general - as threads come and up and threads can go away, wouldn't their numbers vanish from the total as well? Or am I misunderstanding the intent? To be clear: I think we'd want the total number here, as AppDomain does today, not based on some other ephemeral thing that comes and goes. So if the data is available (again: I'm unclear *if* it is), I think we'd want to *include* any dead threads in the aggregate. It *sounds* like that may make the implementation easier as well?

**Comment by Maoni0** [comment]
>as threads come and up and threads can go away, wouldn't their numbers vanish from the total as well

that's why you need to keep track of dead threads. that + the current threads' allocated bytes will give you the total.

**Comment by NickCraver** [comment]
@Maoni0 Okay got it - thanks. My brain was interpreting your comment as *filtering out* dead threads for some reason. On the same page now - I think that tracking of everything alive or dead is the only sensible behavior. I can't think of any other behavior there having a use case (it also wouldn't make the APIs "match" between Core and full framework either).

**Comment by Lexcess** [comment]
While there is definitely some utility in a `GetAllocatedBytesForAllThreads` I can't help but think having a feature to  perform `GetAllocatedBytes` with async operations is more immediately useful. 

@Maoni0 Given the scoped nature of your Stopwatch-esque AllocationTracker proposal would it be possible to get Generation Counts only for the scoped allocations as well as the total byte count?

Lastly assuming you could arbitrarily create a AllocationTracker couldn't you do so at the application/benchmark entry point, and essentially get something close to the use case of getting allocations for all threads anyway?

**Comment by Maoni0** [comment]
@Lexcess 

>would it be possible to get Generation Counts only for the scoped allocations as well as the total byte count

can you please explain why you would want it to get gen counts? you could call GC.CollectionCount yourself obviously but I am guessing there's a reason why that's not sufficient for you. 

>Lastly assuming you could arbitrarily create a AllocationTracker couldn't you do so at the application/benchmark entry point, and essentially get something close to the use case of getting allocations for all threads anyway?

you may or may not - obviously depends how often threads are created not via async. but there is an important difference in perf - this automatic tracking via async locals is more costly than simply reading a number off of each thread object.

**Comment by Lexcess** [comment]
@Maoni0 On the first point it would be more a convenience. If you have a scoping mechanism why not reuse it to try an get as close as possible window on the allocation counts? If we're talking about what I'd ideally want, it's a set of the generation counts for only the scoped objects (not the wider process). From my admittedly limited understanding that would be impossible/non-trivial though.

On the second point I guess it comes down to the use case. I am interested in tracking user code from specific points. This is in order to understand the memory behaviours for a specific algorithm, workflows etc... I'd be willing to trade off some perf for more accurate memory information. If that behaviour had to be opt in for async threads due to perf costs I'd be ok with that.

**Comment by Maoni0** [comment]
> what I'd ideally want, it's a set of the generation counts for only the scoped objects (not the wider process).

there's no "generation counts for only the scoped objects" since GC is per process. GC itself isn't concerned with which scope an object is from aside from recognizing top level objects (ie, objects directly pointed to by roots).  

>  I'd be willing to trade off some perf for more accurate memory information

yeh, that's a reasonable trade off.

**Comment by NinoFloris** [comment]
Still running into this issue with BenchmarkDotNet.
Is there something concrete to push this forward?

**Comment by Maoni0** [comment]
not sure if @NickCraver had time to make more progress on a PR since he last updated?

**Comment by bartonjs** [comment]
@Maoni0 Should this be tracked for 3.0, or Future?

**Comment by NickCraver** [comment]
Just saw pings - we haven't gotten to this at all. I doubt it'll make it high up enough our priority list to dig in further. The issue I hit was I don't see how to do this without adding overhead *everywhere* and I wasn't sure if that was a futile effort for never being worth the performance trade-off.

**Comment by stephentoub** [comment]
This was fixed for 3.0.  AppDomain.MonitoringTotalAllocatedMemorySize works, just delegating to the new GC.GetTotalAllocatedBytes method.
